Add Workflow Title and Annotation sections

hujambo-dunia opened the pull request:
Fixes #18711.

Now, we have two types of views: Run Form version **_(left)_** and the Invocation View one **_(right)_**

![image](https://github.com/user-attachments/assets/96bb64fc-ffdb-4199-b236-138a6a168acb)


<details>
<summary>
<h3>Initial Implementations</h3>
</summary>
Visually, there are numerous ambiguous ways to solve this issue. 

`Version 3` takes into account Marius' comment for simplicity but maintaining consistency with Workflows Landing Page design_.

_v3: Expandable Portlet component with site consistency for Workflows Landing Page_

![demov3_Expandable_Portlet_Site_Consistency_for_Workflows_Landing](https://github.com/user-attachments/assets/98ac709b-da79-4892-9f38-f96040a4cde7)

***

`Version 2` takes into account Ahmed's comment for consistency with Tool Form page header design.

_v2: Expandable Portlet component with site consistency_

![demov2_Expandable_Portlet_Site_Consistency](https://github.com/user-attachments/assets/9ea528b9-fe8b-42d4-bc6b-221c556736e6)

***

`Version 1` is illustrated below as an "expandable portlet" solution. Other ideas/possibilities: "Mouse Hover Overlay-Menu", "Mouse Click Overlay-Menu", "Only Display Description", and so forth. After general solution is agreed, will complete "code clean-up" and test script fix process. Constructive ideas/comments welcome.

_v1: Expandable Portlet component_
![demov1_Expandable_Portlet](https://github.com/user-attachments/assets/ed66c809-2684-4f1a-8e4f-bc8d0ca6646d)

</details>

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

hujambo-dunia committed:
New component for Title and Annotation detail

hujambo-dunia committed:
Added expand/collapse for Annotation detail

hujambo-dunia committed:
Component upgrade, minor styling changes, code cleanup

davelopez commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,71 @@
+    } catch (e) {
+        const error = e as AxiosError<{ err_msg?: string }>;
+        messageVariant.value = "danger";
+        const message = error.response?.data && error.response.data.err_msg;
```
You can `import { errorMessageAsString } from "@/utils/simple-error";` and use it to extract the error message.

davelopez commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,71 @@
+        workflowInfoData.value = await workflowInfoDataPromise;
+    } catch (e) {
+        const error = e as AxiosError<{ err_msg?: string }>;
+        messageVariant.value = "danger";
```
Looks like messageVariant will only be used for "danger" when there is an error, so you can fix the variant to danger and safely remove the redundant ref.

hujambo-dunia replied:
@davelopez very helpful comment - change made.

hujambo-dunia committed:
Cleaner error handling

hujambo-dunia committed:
Added Workflow header 'gray bar' for consistency with Tool Form page design

hujambo-dunia committed:
Linting prettier

mvdbeek commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,70 @@
+const expandAnnotations = ref(true);
+const workflowInfoData = ref(null);
+
+onMounted(() => {
```
onMounted runs after initial rendering and creation of the DOM. You don't need to wait for that.

mvdbeek commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,70 @@
+    loading.value = true;
+
+    try {
+        const workflowInfoDataPromise = getWorkflowInfo(props.model.runData.id);
```
This bypasses all stores and fetches the workflow even if we already have it. Can you use useWorkflowInstance instead ?

An even better alternative is to serialize the workflow annotation for the run request. This would be as simple as
```diff
diff --git a/lib/galaxy/managers/workflows.py b/lib/galaxy/managers/workflows.py
index 0d970de242..5921ad74f4 100644
--- a/lib/galaxy/managers/workflows.py
+++ b/lib/galaxy/managers/workflows.py
@@ -1051,6 +1051,7 @@ class WorkflowContentsManager(UsesAnnotations):
             step_models.append(step_model)
         return {
             "id": trans.app.security.encode_id(stored.id),
+            "annotation": stored.annotation,
             "history_id": trans.app.security.encode_id(history.id) if history else None,
             "name": stored.name,
             "owner": stored.user.username,
```

we only use the run endpoint here in the workflow run component, so if we always need the attribute I don't see a reason not to just serialize that as well.

You would then have this available in the runData object here.

hujambo-dunia replied:
_Follow-up questions related to the [existing _workflow_to_dict_run() method](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/workflows.py#L984) in Workflows.py (eg. maybe some refactoring needed as a separate ticket)._

**(1) Blocker: I am assuming the purpose of @mvdbeek 's code suggestions is to _not_ make an additional database call, correct? However, using the VSCode Debugger to find already-available values, I have not yet been able to find the following variables for the WorkflowInformation.vue component to be fully displayed. Specifically, `creator`, `email_hash`. and `license`. Although, the License may not be appearing because there is no value for it in my dev-qa use-case, but `email_hash` should display (since there is an avatar selected) and `creator` should always have a value. I am assuming that Email Hash can be created from the non-hashed email string parameter: stored.annotations[0].stored_workflow.user.email?** I was able to find/serialize the other values (including Workflow Tags).
(2) Side note: calling `stored.annotation` (singular) fails with the error: 'StoredWorkflow' object has no attribute 'annotation'. The plural value can be accessed using `stored.annotations` ; however, it lacks parameters like the Workflow Tags. Instead, I used `stored.annotations[0].stored_workflow` to get the Workflow Tags - if that is an unreliable value, please let me know.
(3) Side note: using `useWorkflowInstance` pulls up an entirely different Workflow when using the Encoded Workflow ID (that appears in the URL). Tried also with the literal Workflow ID number (as a string) and failed. Perhaps this works only for Invocations since that is where I see `useWorkflowInstance` used in the codebase currently...? The method seemed to be able to used successfully in `WorkflowInvocationHeader.vue` so hopefully @ahmedhamidawan can please advise on this question?
(4) Question: Not a blocker, but I was surprised the following element didn't contain Workflow Annotation data within the same method: `workflow.annotation` returns `null` instead of any actual Workflow Annotation data. If that is not expected, then this line is broken. If "null" is expected, how can this be -is the variable misnamed? [This code block](https://github.com/galaxyproject/galaxy/blame/dev/lib/galaxy/managers/workflows.py#L1015-L1018) is pretty old, but perhaps last editors have some thoughts: @guerler @nsoranzo @dannon ?

ahmedhamidawan replied:
`useWorkflowInstance` takes the stored `workflow_id` as parameter, not the `encoded_workflow_id` we have here from the run data. The stored id is essentially the id for the specific workflow (for the given version; latest in this case) retrieved here:
https://github.com/galaxyproject/galaxy/blob/aa2574343555f10294dc626f991d7b728dfccb5a/lib/galaxy/managers/workflows.py#L916

Once we have that id, the whole `loadAnnotation` method could be replaced by a:
```
const { workflow, loading, error } = useWorkflowInstance(stored_workflow_id);
```

hujambo-dunia replied:
@ahmedhamidawan Thank you so much for this helpful comment. The latest iteration of this (v3) doesn't need this additional data right now, but I may need to know this helpful piece of information in the future.

mvdbeek reviewed (commented):
Really cool, thank you! But please don't embed the published card there. Just a simple expand would be brilliant.

mvdbeek commented on client/src/components/Workflow/Run/WorkflowRunForm.vue:
```diff
@@ -6,16 +6,27 @@
-                :disabled="!canRunOnHistory"
-                :wait="showExecuting"
-                @onClick="onExecute" />
+        <div class="ui-portlet-section">
```
This form is deader than dead. Please don't make changes here, this should only ever be used with maldefined workflows.

ahmedhamidawan replied:
Do you mean the whole `WorkflowRunForm` is dead, as opposed to the `WorkflowRunFormSimple` which is the one we should use and update?

mvdbeek replied:
Yes!

ahmedhamidawan commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,70 @@
+                <span v-localize class="font-weight-bold">About This Workflow</span>
+            </b>
+            <span v-b-tooltip.hover.bottom title="Collapse/Expand" variant="link" size="sm" class="float-right">
+                <FontAwesomeIcon :icon="expandAnnotations ? 'chevron-up' : 'chevron-down'" class="fa-fw" />
```
Minor, but we now use icons like:
```suggestion
                <FontAwesomeIcon :icon="expandAnnotations ? faChevronUp : faChevronDown" fixed-width />
```

... where the two icons are also imported at the top. (We have examples of this in other components)
I think **in the screenshot** the icons are also flipped where: it should be chevron down for an unexpanded one and chevron up for expanded.

hujambo-dunia committed:
Update client/src/components/Workflow/Run/WorkflowRunName.vue

Co-authored-by: Ahmed Hamid Awan <qe66653@umbc.edu>

hujambo-dunia committed:
Reverted changes to client/src/components/Workflow/Run/WorkflowRunForm.vue

hujambo-dunia committed:
Replaced with Workflow-Card component for design consistency with Workflows Landing Page (instead of design consistency with Workflows Annotation History Panel). Added 'show-actions' Boolean parameter to focus user on running the workflow as consistent with their intended behavior from the prior step; when 'show-actions' equals False, all actions are hidden (note: not same terminology as technical 'read-only' as 'show-actions' also prohibits Share/Favorite which are separate from being able to Edit a workflow).

hujambo-dunia commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1059,13 +1076,19 @@ def _workflow_to_dict_run(self, trans, stored, workflow, history=None):
             step_models.append(step_model)
         return {
             "id": trans.app.security.encode_id(stored.id),
+            "annotation": annotations_dict,
```
@mvdbeek Please let me know if it would be even better to serialize these SQLAlchemy objects in `stored.annotations` in some other way, happy to replace.

hujambo-dunia commented on client/src/components/Workflow/Run/WorkflowRunName.vue:
```diff
@@ -0,0 +1,56 @@
+        name: props.model.runData.name,
+        owner: props.model.runData.owner,
+        tags: props.model.runData.annotation.tags.map((t: { user_tname: string }) => t.user_tname),
+        annotations: [props.model.runData.annotation.annotation],
```
@itisAliRH for this line `annotations: [props.model.runData.annotation.annotation],` is ok to only use the first annotation record (ie. `annotation[0]`)? Or are other annotation records need to be included/consolidated somehow?

hujambo-dunia committed:
Backend code formatting

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1059,13 +1079,19 @@ def _workflow_to_dict_run(self, trans, stored, workflow, history=None):
             ]
+            annotations_dict = {
+                "annotation": stored.annotations[0].annotation,
+                "update_time": (stored.annotations[0].stored_workflow.update_time).isoformat(),
```
```suggestion
                "update_time": (stored.update_time).isoformat(),
```
However you're in the step loop, so you're doing this once per step. There's also no guarantee that there is at least one annotation, so `stored.annotations[0].annotation` will fail often.

hujambo-dunia replied:
Thank you! This answers my question.

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1059,13 +1079,19 @@ def _workflow_to_dict_run(self, trans, stored, workflow, history=None):
+            annotations_dict = {
+                "annotation": stored.annotations[0].annotation,
+                "update_time": (stored.annotations[0].stored_workflow.update_time).isoformat(),
+                "tags": self.convert_to_dict_from_many(stored.annotations[0].stored_workflow.tags),
```
```suggestion
                "tags": self.convert_to_dict_from_many(stored.tags),
```
but also we have better ways to serialize tags

mvdbeek commented on lib/galaxy/managers/workflows.py:
```diff
@@ -1059,13 +1079,19 @@ def _workflow_to_dict_run(self, trans, stored, workflow, history=None):
             step_models.append(step_model)
         return {
             "id": trans.app.security.encode_id(stored.id),
+            "annotation": annotations_dict,
```
You're including properties here that are not part of the annotation model.
This is a good attempt that goes in the right direction, I'll try to find some time to rework this.

hujambo-dunia replied:
Sure - any insights & examples are appreciated too.

hujambo-dunia committed:
Update lib/galaxy/managers/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

hujambo-dunia committed:
Update lib/galaxy/managers/workflows.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

hujambo-dunia committed:
Coding convention update

hujambo-dunia committed:
Added reusability & consistency for Workflow headers on: Workflow Invocations Page and Workflow Run Form Page

hujambo-dunia committed:
Added reusability & consistency for Workflow headers on: Workflow Invocations Page and Workflow Run Form Page

ahmedhamidawan committed:
update workflow annotation header with the following changes:

- No `in_panel` prop in route anymore
- In `lib/galaxy/managers`, all we need is the `workflow.id` so on the front end, for the run form, we can call `useWorkflowInstance(workflow_id)`
- Adjust props and variables in the workflow header to cater for all edge cases such as whether it is rendering for run form or invocation view, workflow is owned or not, is published or not etc.
- Minor style improvements

ahmedhamidawan committed:
adjust `WorkflowRunButton` jest for router push

The `BButton :to`, was replaced with a `BButton @click` because the styling in `WorkflowNavigationTitle` worked better with a non `<a>` BButton (using `:to` turns the `<button>` into an `<a>`)

ahmedhamidawan committed:
remove unused imports and props

ahmedhamidawan committed:
in fact do show the history in run form view; just say "Input History"

We removed the switch/view history button from the workflow run form version of the header, but it actually makes sense to keep it there since it lets the user know which history is going to be the input for the workflow (it changes as you change the history).

In the invocation version of course it always made sense because then it is scoped to the history that was used as input for that invocation.

ahmedhamidawan committed:
add a jest each for `WorkflowAnnotation` and `WorkflowNavigationTitle`

ahmedhamidawan committed:
include workflow error in `WorkflowNavigationTitle`

ahmedhamidawan committed:
prevent users from importing workflow over and over

ahmedhamidawan committed:
add space below individual invocation id in `GridInvocation`

ahmedhamidawan committed:
add alert for no metrics, lint

ahmedhamidawan committed:
add badge for workflow run resulting in a new history

ahmedhamidawan committed:
fix jests for workflow header and job metrics
